scanner/src/main/kotlin/experimental/ExperimentalScanner.kt

```diff
@@ -0,0 +1,406 @@
+        scanResults: Map<ScannerWrapper, Map<KnownProvenance, List<ScanResult>>>
+    ): Map<Package, NestedProvenanceScanResult> =
+        packages.associateWith { pkg ->
+            val nestedProvenance = nestedProvenances.getValue(pkg)
+            val provenances = setOf(nestedProvenance.root, *nestedProvenance.subRepositories.values.toTypedArray())
+            val scanResultsByProvenance = provenances.associateWith { provenance ->
+                scanResults.values.flatMap { it[provenance].orEmpty() }
+            }
+            NestedProvenanceScanResult(nestedProvenance, scanResultsByProvenance)
+        }
+
+    private fun scanLocal(
+        provenance: KnownProvenance,
+        scanners: List<LocalScannerWrapper>
+    ): Map<LocalScannerWrapper, ScanResult> {
+        val downloadDir = createOrtTempDir() // TODO: Use provided download dir instead.
+
+        try {
+            provenanceDownloader.download(provenance, downloadDir)
+        } catch (e: DownloadException) {
+            val message = "Could not download provenance $provenance: ${e.collectMessagesAsString()}"
+            log.error { message }
+
+            val summary = ScanSummary(
+                startTime = Instant.now(),
+                endTime = Instant.now(),
+                packageVerificationCode = "",
+                licenseFindings = sortedSetOf(),
+                copyrightFindings = sortedSetOf(),
+                issues = listOf(
+                    OrtIssue(
+                        source = "Downloader",
+                        message = message,
+                        severity = Severity.ERROR
+                    )
+                )
+            )
+
+            return scanners.associateWith { scanner ->
+                ScanResult(
+                    provenance = provenance,
+                    scanner = scanner.details,
+                    summary = summary
+                )
+            }
+        }
+
+        return scanners.associateWith { scanner ->
+            log.info { "Scanning $provenance with local scanner ${scanner.name}." }
+
+            val summary = scanner.scanPath(downloadDir)
+            log.info { "Scan of $provenance with provenance based remote scanner ${scanner.name} finished." }
+
+            ScanResult(provenance, scanner.details, summary)
+        }
+    }
+}
+
+@Suppress("UNCHECKED_CAST")
+private fun <K, V> Map<K, V?>.filterNotNullValues(): Map<K, V> = filterValues { it != null } as Map<K, V>
+
+/**
+ * Split this [ScanResult] into separate results for each [KnownProvenance] contained in the [nestedProvenance] by
+ * matching the paths of findings with the paths in the source tree.
+ */
+fun ScanResult.toNestedProvenanceScanResult(nestedProvenance: NestedProvenance): NestedProvenanceScanResult {
+    val provenanceByPath = listOf(
+        "" to nestedProvenance.root,
+        *nestedProvenance.subRepositories.entries.map { it.key to it.value }.toTypedArray()
```

Comment:
What's the purpose of just mapping keys to values here, isn't this a no-op?

Reply:
It was to transform entires to pairs, but I have replaced it with `toList()` now which does the same.